File: Graph/P207CourseSchedule/CourseSchedule.java
package Graph.P207CourseSchedule;

import java.util.*;

/**
 * @CLassName CourseSchedule
 * @Description TODO
 * @Author cst
 * @Date 2/1/21  1:33 PM
 */
public class CourseSchedule {
    // build graph, and check if the graph has circle
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if (prerequisites.length == 0 || prerequisites[0].length == 0) {
            return true;
        }
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        int[] inDegree = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            graph.put(i, new HashSet<>());
        }
        for (int[] requisite : prerequisites) {
            graph.get(requisite[1]).add(requisite[0]);
            inDegree[requisite[0]]++;
        }
        Set<Integer> visited = new HashSet<>();
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i<inDegree.length; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }
        int count = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int curr = queue.poll();
                count++;
                if (visited.add(curr)) {
                    for (int next : graph.get(curr)) {
                        inDegree[next] -= 1;
                        if (inDegree[next] == 0) {
                            queue.offer(next);
                        }
                    }
                }
            }
        }
        return count == numCourses;
    }
}

File: Graph/P207CourseSchedule/CourseSchedule.py
import unittest
from typing import List
from collections import deque

# using topological sort to check if there ia a circle in the graph
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0] * numCourses
        graph, visited, count = {}, set(), 0
        for i in range(numCourses):
            graph[i] = []
        for req in prerequisites:
            graph[req[1]].append(req[0])
            indegree[req[0]] += 1
        queue = deque()
        
        for i in range(len(indegree)):
            if indegree[i] == 0:
                queue.append(i)

        while queue:
            curr = queue.popleft()
            if curr in visited:
                continue
            visited.add(curr)
            count += 1
            for next in graph[curr]:
                indegree[next] -= 1
                if indegree[next] == 0:
                    queue.append(next)
        return count == numCourses    
        

class UnitTest(unittest.TestCase):
    def test(self):
        numCourses = 2
        prerequisites = [[1,0]]
        self.assertTrue(Solution().canFinish(numCourses, prerequisites))
        
        numCourses = 2
        prerequisites = [[1,0], [0,1]]
        self.assertFalse(Solution().canFinish(numCourses, prerequisites))
        
if __name__ == "__main__":
    unittest.main()